presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -0,0 +1,269 @@
+import com.facebook.presto.server.remotetask.RemoteTaskStats;
+import com.facebook.presto.server.smile.BaseResponse;
+import com.facebook.presto.sql.planner.PlanFragment;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import org.joda.time.DateTime;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.facebook.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static com.facebook.airlift.http.client.Request.Builder.prepareGet;
+import static com.facebook.airlift.http.client.Request.Builder.preparePost;
+import static com.facebook.airlift.http.client.StaticBodyGenerator.createStaticBodyGenerator;
+import static com.facebook.presto.execution.TaskInfo.createInitialTask;
+import static com.facebook.presto.server.RequestHelpers.setContentTypeHeaders;
+import static com.facebook.presto.server.smile.AdaptingJsonResponseHandler.createAdaptingJsonResponseHandler;
+import static com.facebook.presto.spi.StandardErrorCode.NATIVE_EXECUTION_TASK_ERROR;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * NativeExecutionTask provide the abstraction of executing tasks via c++ worker.
+ * The plan is to use it for native execution of Presto on Spark.
+ * The plan and splits is provided at creation of NativeExecutionTask, it doesn't
+ * support adding more splits during execution.
+ */
+public class NativeExecutionTask
+{
+    private static final Logger log = Logger.get(NativeExecutionTask.class);
+    private static final int GET_TASK_INFO_INTERVALS = 100;
+    private final Session session;
+    private final TaskId taskId;
+    private final PlanFragment planFragment;
+    @GuardedBy("this")
+    private final AtomicReference<OutputBuffers> outputBuffers = new AtomicReference<>();
+    private final HttpClient httpClient;
+    private final TableWriteInfo tableWriteInfo;
+    private final RemoteTaskStats stats;
+    private final List<TaskSource> sources;
+    private TaskInfo taskInfo;
+    private final Executor executor;
+    private final ScheduledExecutorService updateScheduledExecutor;
+    @GuardedBy("this")
+    private ScheduledFuture<?> scheduledFuture;
+    private final JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec;
+    private final JsonCodec<PlanFragment> planFragmentCodec;
+    private final JsonCodec<TaskInfo> taskInfoCodec;
+
+    public NativeExecutionTask(
+            Session session,
+            TaskId taskId,
+            String nodeId,
```

Comment:
Can we have nodeId be a constant inside? Or in what scenario we need this nodeId? I might be missing something